File: src/numericalAnalysisAlgorithms/inverseSquareRoot.ts
// TypeScript program for fast inverse square root.

// function to find the inverse square root
function InverseRsqrt(number: number): number {
  const threehalfs: number = 1.5

  let x2: number = number * 0.5
  let y: number = number

  // evil floating point bit level hacking
  let i: number = new Int32Array(new Float32Array([y]).buffer)[0]

  // value is pre-assumed
  i = 0x5f3759df - (i >> 1)
  y = new Float32Array(new Int32Array([i]).buffer)[0]

  // 1st iteration
  y = y * (threehalfs - x2 * y * y)

  // 2nd iteration, this can be removed
  // y = y * (threehalfs - x2 * y * y);

  return y
}

// Export functions
export { InverseRsqrt }

// Test
// const input: number = 4
// const result: number = InverseRsqrt(input)
// console.log(result) // Output: 0.49915357479239103

File: src/numericalAnalysisAlgorithms/inverseSquareRoot.c
// C program for fast inverse square root.
#include <stdio.h>

float inverse_rsqrt(float number)
{
  long i;
  float x2, y;
  const float threehalfs = 1.5F;

  x2 = number * 0.5F;
  y = number;
  i = *(long *)&y;           // evil floating point bit level hacking
  i = 0x5f3759df - (i >> 1); // what the fuck?
  y = *(float *)&i;
  y = y * (threehalfs - (x2 * y * y)); // 1st iteration
  // y  = y * ( threehalfs - ( x2 * y * y ) );   // 2nd iteration, this can be removed

  return y;
}

// driver code
int main()
{
  int n = 0;
  printf("Enter number: ");
  scanf("%d", &n);
  float f = inverse_rsqrt(n);
  printf("Entered number: %d", n);
  printf(" inverse square root is: %f", f);
  return 0;
}

/*
src/numericalAnalysisAlgorithms> gcc ./inverseSquareRoot.c
src/numericalAnalysisAlgorithms> a.out
*/
